### PR TITLE
Feature/callback registration

### PR DESCRIPTION
**Pull request by mibrasap**
This feature enables the automated callback registration at the EFGS during startup.
- The env variable CALLBACK_REGISTER_ON_STARTUP determines if the registration should take place
- The route of the callback service to be registered must be placed in the env variable CALLBACK_ENDPOINT_URL
- The federation-gateway.ssl variables are required to establish the connection to the EFGS

### CODE REVIEW

**Review by ioangut — DISMISSED** [review]

File: services/callback/src/main/java/app/coronawarn/server/services/callback/CallbackUtils.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package app.coronawarn.server.services.callback;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import org.bouncycastle.util.encoders.Hex;
+
+public class CallbackUtils {
```
[the remark this answers is not in the document]
Comment (mibrasap) [reply services/callback/src/main/java/app/coronawarn/server/services/callback/CallbackUtils.java]: Thanks for the hint, it's changed to `HashingUtils`

**Review by ioangut — DISMISSED** [review]

**Review by EugenM-SAP — DISMISSED** [review]

File: common/federation/src/main/java/app/coronawarn/server/common/federation/client/FederationGatewayClient.java [hunk 2]
```diff
@@ -25,6 +26,19 @@
...
 public interface FederationGatewayClient {
 
+  @GetMapping(value = "/diagnosiskeys/callback",
+      headers = {"Accept=application/json; version=1.0",
+          "X-SSL-Client-SHA256=${federation-gateway.ssl.certificate-sha}",
+          "X-SSL-Client-DN=${federation-gateway.ssl.certificate-dn}"})
+  ResponseEntity<List<RegistrationResponse>> getCallbackRegistrations();
+
+  @PutMapping(value = "/diagnosiskeys/callback/{id}?url={url}",
+      headers = {"Accept=application/json; version=1.0",
```
Comment (hilmarf) [inline common/federation/src/main/java/app/coronawarn/server/common/federation/client/FederationGatewayClient.java:36]: version for 'accept'!?! didn't know that this is possible/necessary

File: services/callback/src/main/java/app/coronawarn/server/services/callback/HashingUtils.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+   *
+   * @param subject the string to compute the hash
+   * @return the SHA-256 hash of the string.
+   */
+  public static String computeSha256Hash(String subject) {
+    MessageDigest digest = null;
+    try {
+      digest = MessageDigest.getInstance(HASHING_ALGORITHM);
+    } catch (NoSuchAlgorithmException e) {
+      throw new HashAlgorithmNotFoundException("Hashing algorithm '" + HASHING_ALGORITHM + "' not found.");
```
Comment (hilmarf) [inline services/callback/src/main/java/app/coronawarn/server/services/callback/HashingUtils.java:27]: is there a reason, why we silently drop the complete stacktrace and ignore the root-cause?

```suggestion
      throw new HashAlgorithmNotFoundException("Hashing algorithm '" + HASHING_ALGORITHM + "' not found.", e);
```

File: services/callback/src/main/java/app/coronawarn/server/services/callback/HashAlgorithmNotFoundException.java [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package app.coronawarn.server.services.callback;
+
+public class HashAlgorithmNotFoundException extends RuntimeException {
+
+  public HashAlgorithmNotFoundException(String msg) {
```
Comment (hilmarf) [inline services/callback/src/main/java/app/coronawarn/server/services/callback/HashAlgorithmNotFoundException.java:5]: why not including the root cause?
```suggestion
  public HashAlgorithmNotFoundException(String msg, Throwable rootCause) {
```

File: services/callback/src/main/java/app/coronawarn/server/services/callback/registration/RegistrationRunner.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import org.springframework.boot.ApplicationRunner;
+import org.springframework.core.annotation.Order;
+import org.springframework.stereotype.Component;
+
+/**
+ * This runner registers the callback service at the EFGS.
+ */
+@Component
+@Order(1)
+public class RegistrationRunner implements ApplicationRunner {
```
Comment (hilmarf) [inline services/callback/src/main/java/app/coronawarn/server/services/callback/registration/RegistrationRunner.java:20]: `ApplicationRunner` do work together with long running services? cool!

File: services/callback/src/main/java/app/coronawarn/server/services/callback/registration/RegistrationRunner.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+  public RegistrationRunner(CallbackServiceConfig serviceConfig, FederationGatewayClient federationGatewayClient) {
+    this.serviceConfig = serviceConfig;
+    this.federationGatewayClient = federationGatewayClient;
+  }
+
+  @Override
+  public void run(ApplicationArguments args) {
+    if (!serviceConfig.isRegisterOnStartup()) {
+      logger.info("Callback registration on startup was disabled.");
```
Comment (hilmarf) [inline services/callback/src/main/java/app/coronawarn/server/services/callback/registration/RegistrationRunner.java:34]:
```suggestion
      logger.info("Callback registration on startup is disabled. See 'CALLBACK_REGISTER_ON_STARTUP'");
```

File: services/callback/src/main/resources/application.yaml [hunk 7]
```diff
@@ -10,6 +10,17 @@ logging:
 services:
   callback:
     efgs-cert-cn: ${EFGS_CERT_CN}
+    register-on-startup: ${CALLBACK_REGISTER_ON_STARTUP:false}
+    endpoint-url: ${CALLBACK_ENDPOINT_URL}
```
Comment (hilmarf) [inline services/callback/src/main/resources/application.yaml:14]:
```suggestion
    register-on-startup: ${CALLBACK_REGISTER_ON_STARTUP:true}
    endpoint-url: ${CALLBACK_ENDPOINT_URL}
```

File: services/callback/src/main/java/app/coronawarn/server/services/callback/registration/RegistrationRunner.java (hunk 5 above)
Comment (mibrasap) [reply services/callback/src/main/java/app/coronawarn/server/services/callback/registration/RegistrationRunner.java:20]: Yes, their original purpose is to execute logic right before the start of the application. Quite a cool concept which we use to run our actual logic once after start.

**Review by ioangut — APPROVED** [review]

**Review by hilmarf — APPROVED** [review]

**Pull request merged**